README.md
push
====

A Zandronum multiplayer mod about throwing people off of buildings. Wow.


changelog.txt
FROM V0.7 TO V0.8:
==================
- Players are no longer capable of drowning.
- The death camera now works.
- Players are now no longer capable of dying after they've been killed.
- Flying no longer triggers both doublejump or surfing counter.


FROM V0.6 TO V0.7:
==================
GAMEPLAY:
- Surfing is no longer as obscenely exploitable as before--when a player bounces repeatedly in mid-air without landing, he will eventually stop bouncing and will just plummet.
- Players are now always bright, for ease of visibility at longer ranges.
- Drones are now faster, fire faster, and their shots speed up. There are also more drone spawns on certain maps.
 ^- Push08
 ^- Push19
 ^- Push10
 ^- Push11
 ^- Push15
 ^- Push18
 ^- Push21
 ^- Push22
 ^- Push24
 ^- Push31
 ^- Push45
- Changing/swapping maps while in non-DM modes no longer renders the players invulnerable.
- Sudden death drones now only show up in Last Man Standing/Team Last Man Standing.
- Push is now compatible with and playable in ZDoom/GZDoom! I don't know WHY you'd want to, but there you have it!
- Respawn protection now is taken away as soon as players fire, so players are now able to rocketjump out of the gate.
- p_playerdrones has been added as a cvar. When enabled, when the player is on his final life (and sv_maxlives is above one), he will be turned into a drone to chase down the remaining players as he sees fit.
- p_nohazards has been added as a cvar. When enabled, all randomized dangers or map-changing hazards will be disabled. This affects maps:
 ^- Push04
 ^- Push16
 ^- Push17
 ^- Push27
 ^- Push32
 ^- Push34
 ^- Push42
 ^- Push44
 ^- Push60
 ^- Push61
 ^- Push63
 ^- Push69
- Push weapons no longer support autoaim.
- The push laser was made one tic slower, but bolts now travel over 1.5x as fast. This should easily help its role in long-ranged nudging of opponents trying to recover.
- If sv_nomedals is not turned on, it will automatically be turned on.

MAPS:
- Push11 now has both drone spawns and projectile blocking boundaries.
- Push19's been given the Presidential Treatment:
 ^ More efficient method of keeping players off the counter (an invisible 3D floor) end
 ^ Fixed the alignment of some pyramid light textures
 ^ Blue ring platforms are now fully lit and 125% wider
 ^ Raised ceilings above some 3D floors slightly, so they appear on the automap
 ^ Filled the large window in with glass.  The center pane remains open
 ^ Fountains now launch players up
- Push20's boundaries have been shortened by 2056 units.
- Push27 now has a drone spawn.
- Absolutely nothing was changed about Push28.
- Push30 no longer can teleport lasers into infinity.
- Push32 now has darker borders around the circles to indicate position.
- Push33's center mountains have been removed, and the mountains have been moved to the outer portion of the map. To address camping, there is now an ocean drift on the outside of the map--get thrown out there, and you'll start drifting away!
- Push34's platforms have been trimmed and the belts sped up.
- Push34 has many more spawnpoints now.
- Push35 no longer has player spawns outside of the play boundary.
- Push38's layout has been squeezed in a slight bit, for easier mobility.
- Push42's gravity now resets every time the map restarts.
- Push42's gravity will now always adjust by at least 10%.
- Push43's killfloor has had its depth increased, so it's no longer as easy to trip into it.
- Push46's skybox is no longer obscenely fuckhuge filesize.
- Push47's camp spots have been removed.
- Push47's miscolored water has been colored properly.
- Push49's boundaries have been shortened and the main layout has been expanded.
- Push49 no longer has the water bobbing up and down.

- A new map, Ichikishimahime Says Keep Your Water Clean by Kurashiki, has been added at Push50.
- A new map, Coffee Break by PresidentPeople, has been added at Push51.
- A new map, Ship of Fools by UniversalQuantifier, has been added at Push52.
- A new map, Samhara Trench by IdeIdoom, has been added at Push53.
- A new map, Dwango 2015 by Gardevoir, has been added at Push54.
- A new map, Burn Baby Burn by Infirnex, has been added at Push55.
- A new map, Jazzy Nostalgia by IdeIdoom, has been added at Push56.
- A new map, Bullseye by Sgt. Shivers, has been added at Push57.
- A new map, Heifong Virtual Arena by Marty Kirra, has been added at Push58.
- A new map, Push-Mortem by Untitled, has been added at Push59.
- A new map, Castle of Griefing by PresidentPeople, has been added at Push60.
- A new map, Serpent Riders' Playhouse by Gothic, has been added at Push61.
- A new map, Island Retreat by PresidentPeople, has been added at Push62.
- A new map, Wild is the Wind by Infirnex, has been added at Push63.
- A new map, Acid Killer by UniversalQuantifier, has been added at Push64.
- A new map, 7th Hell by Gardevoir, has been added at Push65.
- A new map, Gimmick! by Deadon, has been added at Push66.
- A new map, The Strongest Spirit by Guardsoul, has been added at Push67.
- A new map, The Longest Push by Infirnex, has been added at Push68.
- A new map, Gardie's Crazy Game Show! by Gardevoir and Guardsoul, has been added at Push69.
- A new map, Peaceful Hexagon Realm by Yholl, has been added at Push70.


POLISH:
- Sgt. Shivers has provided the best titlepic.
- All SNDINFO entries have been separated by map.
- Push19 had had a massive wipe of polish across the entire board, with more effects, prettier textures, and some new additions.
- Push24 now has over double the banners. Thanks to everyone for submitting!
- Push31's HOMs behind the banner has been removed.
- Push34 now has rain and lightning.
- Push30 now has a Startropics enemy wandering about.
- Push38 has been given its proper skybox.
- Several misaligned sectors in Push43 have been re-aligned. Not all of them, but quite a few.
- Push48 has been given more even lighting across the board.

pk3/acs/push-copy.c
#include "zcommon.acs"
#library "push"

world int 0:MaxJumpCount;
global int 58:TimeExceeded;

int IsServer;
int FontX;
int FontY;
int FontTim;
str FontZ[2] ={"0","1"};
int FontCL[3] ={CR_DARKGREEN,CR_GREEN,CR_OLIVE};

#include "commonfuncs.h"
#include "push_const.h"
#include "push_funcs.h"
#include "push_text.h"
#include "push_jump.h"
#include "push_dm.h"

int lastShotBy[PLAYERMAX];
int deathAssistedBy[PLAYERMAX];

int PlayerTIDs[PLAYERMAX];

script PUSH_SERVER OPEN
{
    IsServer = 1;
    int cjumps, oldcjumps;
    int s = unusedTID(37000, 47000);
    int players = PlayerCount();
    int drones = 0;

    /*if (Spawn("Doomsphere",0,0,0,s)) // Attempt to spawn a Doomsphere, a Zandronum-exclusive item.
    {
        Thing_Remove(s);
        SetCVar("p_runninginzandro",1);
        SetCVar("p_runninginzdoom",0);
    }*/
    // This didn't work because the Spawn would choose to not work at various times, because of collision.
    // If there is no room for an actor to exist, it simply won't spawn.
    // To get around this, instead I'm doing it the opposite way, using SpawnForced to spawn a ZDoom-exclusive item.
    // This is not a future-proof method at all; if Zandronum catches up with ZDoom (hahahaha), it'll spawn this item as well and read as ZDoom.
    // If we ever reach that point and I am either dead or have moved on from Doom, simply replace the script with the uncommented above script and change it to SpawnForced.
    if (SpawnForced("SpeakerIcon",0,0,0,s,0))
    { // Other ZDoom-exclusive items include PowerBuddha (not in GZDoom 1.9), CajunTrace, CajunBodyNode, ArtiPoisonBagShooter, ArtiPoisonBagGiver, and AmbientSoundNoGravity.
        Thing_Remove(s);
        SetCVar("p_runninginzandro",0);
        SetCVar("p_runninginzdoom",1);
    }
    else
    {
        SetCVar("p_runninginzandro",1);
        SetCVar("p_runninginzdoom",0);
    }

    while (1)
    {

    SetAirControl(0.01);

    if (!GetCvar("compat_clientssendfullbuttoninfo") && GetCvar("p_runninginzandro") == 1)
        { ConsoleCommand("set compat_clientssendfullbuttoninfo 1"); }

    if (GetCvar("lmsallowedweapons") != 0 && GetCvar("p_runninginzandro") == 1)
        { ConsoleCommand("set lmsallowedweapons 0"); }

    if (!GetCvar("sv_nocrouch") && GetCvar("p_runninginzandro") == 1)
        { ConsoleCommand("set sv_nocrouch 1"); }

        oldcjumps = cjumps;
        cjumps = GetCVar("p_jumpcount");
        if (cjumps != oldcjumps) { MaxJumpCount = cjumps; }

    for (int i = PlayerTIDs[PLAYERMAX]; i < PlayerTIDs[PLAYERMAX] + PLAYERMAX; i++)
      if (CheckActorInventory(i, "IsADrone"))
        drones++;
    if (players - drones <= 1)
    {
        PrintBold(s:"One player left.");
    }

        delay(1);
    }
}

Script PUSH_SUDDENDEATH OPEN
{
    // Probably could compress this into the above script, but I'd rather not take a chance.
    int TimeUntilSuddenDeath = 1;

    delay(5);

    if (GetCvar("p_suddendeath") == 1 && isLMS()) // Doesn't show up in other game modes because there's no real need to end a match fast, is there?
    {
        while(GetCvar("p_timelimit") >= TimeUntilSuddenDeath)
        {
            delay(2100);
            //PrintBold(s:"2100 tics have passed.");
            TimeUntilSuddenDeath++;
        }
        if (GetCvar("p_timelimit") <= 0) { delay(2100); }

        //PrintBold(s:"Preparing for sudden death.");

        if (RealPlayerCount() > 1 && GetCvar("p_suddendeath") == 1)
        {
            AmbientSound("suddendeath/alarm",127);
            SetFont("smallfont");
            HudMessageBold(s:"WARNING: INVALID ACCESS DETECTED";HUDMSG_TYPEON,0,CR_RED,0.5,0.2,3.5,0.05);

            while (1)
            {
                SpawnSpot("TeleportFog",130);
                SpawnSpot("SuddenDeathDrone",130,1313);
                delay(350);
                if (RealPlayerCount() == 0)
                {
                    Thing_Remove(1313);
                    delay(1);
                    terminate;
                }
            }
        }
    }
}

script PUSH_ENTER ENTER
{
    int Buttons;
    int Angle;
    int DropSomeWeight;
    int tid;
    int pln = PlayerNumber();

    TakeInventory("LightAsAFeather",1);
    TakeInventory("EmergencyDodgeDone",1);
    TakeInventory("DrawingToolOn",1);
    TakeInventory("DrawingToolReady",1);
    GiveInventory("ImAlive",1);
    //SetPlayerProperty(0,1,PROP_BUDDHA);
    GiveInventory("SetUnshootable",1);
    GiveInventory("PowerRespawnProtection",1);

    while (1)
    {
      if (CheckInventory("IsADrone") == 0)
      {
        TakeInventory("Fist",1);
        TakeInventory("Pistol",1);
        TakeInventory("Shotgun",1);
        TakeInventory("SuperShotgun",1);
        TakeInventory("Chaingun",1);
        TakeInventory("PlasmaRifle",1);
        TakeInventory("BFG9000",1);
        TakeInventory("Chainsaw",1);
        // Just in case lms_allowedweapons 0 doesn't take or just in case someone decides to give weapons or whatever.
        TakeInventory("Drone Gun",1);

        tid = defaultTID(-1);
        PlayerTIDs[pln] = tid;

        GiveInventory("500Health",500);

        if (GetCvar("p_punchdrunk") == 1)
          { TakeInventory("Push Gun",1); }
        else
          { GiveInventory("Push Gun",1); }
        if (GetCvar("p_triggerhappy") == 1)
          { TakeInventory("Force Gauntlet",1); }
        else
          { GiveInventory("Force Gauntlet",1); }

        if (DropSomeWeight == 70 && CheckInventory("LightAsAFeather") == 0)
        {
            GiveInventory("SetShootable",1);
            GiveInventory("LightAsAFeather",1);
        }
        DropSomeWeight++;

        Buttons = GetPlayerInput(-1, INPUT_BUTTONS);
        Angle = GetActorAngle(0) >> 8;

        if (CheckInventory("DoomedGuy") > 0) { GiveInventory("DunkedSmokeActivator",1); }

        if (GetActorZ(0) - GetActorFloorZ(0) == 0)
        {
            GiveInventory("OnTheGround", 1);
            TakeInventory("EmergencyDodgeDone",1);
            TakeInventory("DoomedGuy",70);
            TakeInventory("SurfingCounter",15);
        }
        else
        {
            TakeInventory("OnTheGround", 1);
        }

        if (keyDown(BT_USER1)) { GiveInventory("DrawingToolOn",1); }

        if (GetCvar("p_runninginzandro") == 1)
        {
          if (GetCvar("sv_maxlives") > 1 && GetCvar("p_playerdrones") == 1 && GetPlayerLivesLeft(pln) == 0 && PlayerIsSpectator(pln) == 0 && isLMS() && PlayerCount() >= 2) // Bloody hell that's a lot of checks.
          { 
            MorphActor(0, "DronePlayer", "", 0x7FFFFFFF, 194, "TeleportFog", "TeleportFog");
            TakeInventory("Push Gun",1);
            TakeInventory("Force Gauntlet",1);
            GiveInventory("IsADrone",1);
          }
        }
        if (isDead(0)) { terminate; }

        delay(1);
      }
      else if (CheckInventory("IsADrone") == 1)
      {
        GiveInventory("PowerPermaFlight",1);
        if (isDead(0)) { terminate; }
        delay(1);
      }
    }
}

script PUSH_RESPAWN RESPAWN
{
    ACS_ExecuteAlways(532,0,0,0);
}

script PUSH_DEATH DEATH
{
    //int pln = PlayerNumber();
    //Log(n:pln, s:" was thrown off by ", n:lastShotBy[pln]);

    TakeInventory("ImAlive",1);
    TakeInventory("LightAsAFeather",1);
    TakeInventory("DrawingToolOn",1);
    TakeInventory("DrawingToolReady",1);
    Thing_ChangeTID(0,0);
}

script PUSH_WALLHACK (int which)
{
    int i, time, tid;
    int pln = PlayerNumber();

    switch (which)
    {
      case 1:
        GiveInventory("WallHackVision",1);
        GiveInventory("CannotIntoWallhack", 1);

        while (1)
        {
            for (i = 0; i < PLAYERMAX; i++)
            {
                if (i == pln || !PlayerInGame(i)) { continue; }
                tid = PlayerTIDs[i];
                if (isDead(tid)) { continue; }

                HudMessageOnActor(tid, 0x7FFFFFFF, "PLAYMARK", "A", tid);
            }

            Delay(2);

            if (!CheckInventory("WallhackVision")) { break; }
            TakeInventory("CannotIntoWallhack", 1);

            Delay(2);
            if (!CheckInventory("WallhackVision")) { break; }
        }
        break;

      case 2:
        TakeInventory("WallHackVision",1);
        GiveInventory("CannotIntoWallhack",1);
        Delay(15);
        TakeInventory("CannotIntoWallhack",1);
        break;
    }
}

// The below was made by Kyle873, whom is 7H3 1337357 |-|4(|<3R 0N 7H3 PU5H|\|37

script PUSH_DUNKED (void)
{
    int Time = 52;

    while (Time > 0)
    {
        SetHudSize(320, 240, false);

        // Scanlines
        if ((Timer() % 3) == 0)
            DrawScanlines();

        // Binary
        if ((Timer() % 3) == 0)
            DrawBinary();

        SetFont("bigfont");
        FontX = random(0.00,320.00);
        FontY = random(0.00,240.00);
        hudmessage (s:FontZ[random(0,1)],s:"\n";
        1, 0,FontCL[random(0,2)], FontX, FontY, 0.1);
        delay(1);

        Time--;
        Delay(1);
    }
}

/*<04:55:27> "Kyle873": 
int players = PlayerCount();
int drones = 0;
for (int i = PlayerTIDs; i < PlayerTIDs + MAX_PLAYERS; i++)
    if (CheckActorInventory(i, "IsADrone"))
        drones++;
if (players - drones <= 1)
{
    // We've only got one player left
}

<04:56:25> "Kyle873": 
if (players - drones <= 1)
{
    for (int i = PlayerTIDs; i < PlayerTIDs + MAX_PLAYERS; i++)
        Thing_Damage(i, 1000000;
}

pk3/acs/push_funcs.h
function void hudmessageonactor(int tid, int range, str sprite, str text, int id) // By Caligari 87
{
    int dist, ang, vang, pitch, x, y;
    int HUDX = 1280;
    int HUDY = 960;
    int offset = 0;

    if(sprite != -1)
    {
        setfont(sprite);
        text = "A";
        offset = 0.1;
    }

    sethudsize(HUDX, HUDY, 1);
    x = getactorx(tid) - getactorx(0);
    y = getactory(tid) - getactory(0); 

    vang = vectorangle(x,y);
    ang = (vang - GetActorAngle(0) + 1.0) % 1.0;

    if(((vang+0.125)%0.5) > 0.25) dist = fixeddiv(y, sin(vang));
    else dist = fixeddiv(x, cos(vang));

    if ((ang < 0.2 || ang > 0.8) && dist < range)
    {
        pitch = vectorangle(dist, getactorz(tid) - (getactorz(0) + 41.0));
        pitch = (pitch + GetActorPitch(0) + 1.0) % 1.0;

        x = HUDX/2 - ((HUDX/2) * sin(ang) / cos(ang));
        y = HUDY/2 - ((HUDX/2) * sin(pitch) / cos(pitch));

        if(sprite != -1)
            setfont(sprite);
        hudmessage(s:text; HUDMSG_FADEOUT, id, CR_UNTRANSLATED, (x<<16)+offset, ((y<<16)+offset)-32.0, 0.25, 0.5);
    }
    else { hudmessage(s:" "; HUDMSG_PLAIN, id, CR_UNTRANSLATED, 0, 0, 0.1); }
}

function void DrawScanlines(void)
{
    int Y = Random(0, 240) << 16;
    int Alpha = FixedDiv(Random(25, 75), 100);
    
    // Scanlines
    SetFont("Line");
    HudMessage(s:"A"; HUDMSG_FADEOUT, 0, CR_WHITE, 0.0, Y, 0.05, 1.0, Alpha);
}

function void DrawBinary(void)
{
    int X = Random(0, 320) << 16;
    int Y = Random(0, 240 / 6) << 16;
    int FadeInTime = 0.1;
    int FadeOutTime = 0.25;
    int Rows = Random(10, 30);
    str Binary;
    
    for (int i = 0; i < Rows; i++)
    {
        if (Random(0, 1))
            Binary = "1";
        else
            Binary = "0";
        
        SetFont("SMALLFONT");
        HudMessage(s:Binary; HUDMSG_FADEINOUT, 0, CR_GREEN, X, Y, 0.05, FadeInTime, FadeOutTime, 0.75);
        
        FadeInTime += 0.1;
        FadeOutTime += 0.1;
        Y += 8.0;
    }
    
    // OH NO
    if (!Random(0, 100))
    {
        SetFont("BIGFONT");
        HudMessage(s:"DIVIDE BY ZERO"; HUDMSG_FADEOUT, 0, CR_GREEN, X, Y - 100.0, 0.05, 3.0, 0.75);
    }
}

pk3/acs/push.c
#include "zcommon.acs"
#library "push"

world int 0:MaxJumpCount;
global int 58:TimeExceeded;
global int 59:DieDieTimeToDie;
global int 60:OhShitItsTheFeds;

int IsServer;
int FontX;
int FontY;
int FontTim;
str FontZ[2] ={"0","1"};
int FontCL[3] ={CR_DARKGREEN,CR_GREEN,CR_OLIVE};

#include "commonfuncs.h"
#include "push_const.h"
#include "push_funcs.h"
#include "push_text.h"
#include "push_jump.h"
#include "push_dm.h"

int lastShotBy[PLAYERMAX];
int deathAssistedBy[PLAYERMAX];

int PlayerTIDs[PLAYERMAX];

script PUSH_WOWBYE (int pnum) DISCONNECT
{
  PlayerTIDs[pnum] = 0;
}

script PUSH_SERVER OPEN
{
    IsServer = 1;
    int cjumps, oldcjumps;
    int s = unusedTID(37000, 47000);
    int players;
    int playercnt; // NOT playercunt
    int _PTID;
    int i;


    /*if (Spawn("Doomsphere",0,0,0,s)) // Attempt to spawn a Doomsphere, a Zandronum-exclusive item.
    {
        Thing_Remove(s);
        SetCVar("p_runninginzandro",1);
        SetCVar("p_runninginzdoom",0);
    }*/
    // This didn't work because the Spawn would choose to not work at various times, because of collision.
    // If there is no room for an actor to exist, it simply won't spawn.
    // To get around this, instead I'm doing it the opposite way, using SpawnForced to spawn a ZDoom-exclusive item.
    // This is not a future-proof method at all; if Zandronum catches up with ZDoom (hahahaha), it'll spawn this item as well and read as ZDoom.
    // If we ever reach that point and I am either dead or have moved on from Doom, simply replace the script with the uncommented above script and change it to SpawnForced.
    if (SpawnForced("SpeakerIcon",0,0,0,s,0))
    { // Other ZDoom-exclusive items include PowerBuddha (not in GZDoom 1.9), CajunTrace, CajunBodyNode, ArtiPoisonBagShooter, ArtiPoisonBagGiver, and AmbientSoundNoGravity.
        Thing_Remove(s);
        SetCVar("p_runninginzandro",0);
        SetCVar("p_runninginzdoom",1);
    }
    else
    {
        SetCVar("p_runninginzandro",1);
        SetCVar("p_runninginzdoom",0);
    }

    DieDieTimeToDie = 0;

    while (1)
    {

    SetAirControl(0.01);

    if (GetCvar("compat_clientssendfullbuttoninfo") != 1 && GetCvar("p_runninginzandro") == 1)
        { ConsoleCommand("set compat_clientssendfullbuttoninfo 1"); }

    if (GetCvar("lmsallowedweapons") != 0 && GetCvar("p_runninginzandro") == 1)
        { ConsoleCommand("set lmsallowedweapons 0"); }

    if (GetCvar("sv_nocrouch") != 1 && GetCvar("p_runninginzandro") == 1)
        { ConsoleCommand("set sv_nocrouch 1"); }

    if (GetCvar("sv_nomedals") != 1 && GetCvar("p_runninginzandro") == 1)
        { ConsoleCommand("set sv_nomedals 1"); }

        oldcjumps = cjumps;
        cjumps = GetCVar("p_jumpcount");
        if (cjumps != oldcjumps) { MaxJumpCount = cjumps; }
        
        // cod by murb
        // donut steel
        
        if(!DieDieTimeToDie)
        {
          playercnt = PlayerCount();
          players = 0;
          
          for (i = 0; i < PLAYERMAX; i++)
          {
            _PTID = PlayerTIDs[i];
            
            if (_PTID != 0) // don't condense this into one if!
            {
              if (CheckActorInventory(_PTID, "IsADrone") == 0)
              {
                //PrintBold(s:"TID (not a drone): ", d:_PTID);
                players++;
              }
            }
          }
          
          //PrintBold(s:"Players: ", d:players, s:", Player Count: ", d:playercnt);
          
          if (players < 2 && playercnt > 1) // if less than two people are left and more than 1 person in-game...
          {
            // kill all the drones!
            //PrintBold(s:"One player left. Commencing slaughterstructionnihilator.");
            DieDieTimeToDie = 1;
          }
        }
        // ebn murb cod
        
        delay(1);
    }
}

Script PUSH_SUDDENDEATH OPEN
{
    // Probably could compress this into the above script, but I'd rather not take a chance.
    int TimeUntilSuddenDeath = 1;

    OhShitItsTheFeds = 0;

    delay(5);

    if (GetCvar("p_suddendeath") == 1 && isLMS()) // Doesn't show up in other game modes because there's no real need to end a match fast, is there?
    {
        while(GetCvar("p_timelimit") >= TimeUntilSuddenDeath)
        {
            delay(2100);
            //PrintBold(s:"2100 tics have passed.");
            TimeUntilSuddenDeath++;
        }
        if (GetCvar("p_timelimit") <= 0) { delay(2100); }

        //PrintBold(s:"Preparing for sudden death.");

        if (RealPlayerCount() > 1 && GetCvar("p_suddendeath") == 1)
        {
            AmbientSound("suddendeath/alarm",127);
            SetFont("smallfont");
            HudMessageBold(s:"WARNING: INVALID ACCESS DETECTED";HUDMSG_TYPEON,0,CR_RED,0.5,0.2,3.5,0.05);
            OhShitItsTheFeds = 1;
            
            while (1)
            {
                SpawnSpot("TeleportFog",130);
                SpawnSpot("SuddenDeathDrone",130,1313);
                delay(350);
                if (RealPlayerCount() == 0)
                {
                    Thing_Remove(1313);
                    delay(1);
                    terminate;
                }
            }
        }
    }
}

script PUSH_ENTER ENTER
{
    int Buttons;
    int Angle;
    int DropSomeWeight;
    int tid;
    int newtid;
    int pln = PlayerNumber();

    TakeInventory("LightAsAFeather",1);
    TakeInventory("EmergencyDodgeDone",1);
    TakeInventory("DrawingToolOn",1);
    TakeInventory("DrawingToolReady",1);
    GiveInventory("ImAlive",1);
    //SetPlayerProperty(0,1,PROP_BUDDHA);
    GiveInventory("SetUnshootable",1);
    GiveInventory("PowerRespawnProtection",1);

    while (1)
    {
      if (CheckInventory("IsADrone") == 0)
      {
        TakeInventory("Fist",1);
        TakeInventory("Pistol",1);
        TakeInventory("Shotgun",1);
        TakeInventory("SuperShotgun",1);
        TakeInventory("Chaingun",1);
        TakeInventory("PlasmaRifle",1);
        TakeInventory("BFG9000",1);
        TakeInventory("Chainsaw",1);
        // Just in case lms_allowedweapons 0 doesn't take or just in case someone decides to give weapons or whatever.
        TakeInventory("Drone Gun",1);

        tid = defaultTID(-1);
        PlayerTIDs[pln] = tid;

        if (GetCvar("p_punchdrunk") == 1)
          { TakeInventory("Push Gun",1); }
        else
          { GiveInventory("Push Gun",1); }
        if (GetCvar("p_triggerhappy") == 1)
          { TakeInventory("Force Gauntlet",1); }
        else
          { GiveInventory("Force Gauntlet",1); }

        if (DropSomeWeight == 70 && CheckInventory("LightAsAFeather") == 0)
        {
            GiveInventory("SetShootable",1);
            GiveInventory("LightAsAFeather",1);
        }
        DropSomeWeight++;

        Buttons = GetPlayerInput(-1, INPUT_BUTTONS);
        Angle = GetActorAngle(0) >> 8;

        if (CheckInventory("DoomedGuy") > 0) { GiveInventory("DunkedSmokeActivator",1); }

        if (GetActorZ(0) - GetActorFloorZ(0) == 0)
        {
            GiveInventory("OnTheGround", 1);
            TakeInventory("EmergencyDodgeDone",1);
            TakeInventory("DoomedGuy",70);
            TakeInventory("SurfingCounter",15);
        }
        else
        {
            TakeInventory("OnTheGround", 1);
        }

        if (keyDown(BT_USER1)) { GiveInventory("DrawingToolOn",1); }

        if (GetCvar("p_runninginzandro") == 1)
        {
          if (GetCvar("sv_maxlives") > 1 && GetCvar("p_playerdrones") == 1 && GetPlayerLivesLeft(pln) == 0 && PlayerIsSpectator(pln) == 0 && isLMS() && PlayerCount() >= 2) // Bloody hell that's a lot of checks.
          {
 			newTID = unusedTID(23000, 55000);
			Spawn("TranslationHolder",GetActorX(0),GetActorY(0),GetActorZ(0)+8.0,newTID);
			Thing_SetTranslation(newTID, -1);
            MorphActor(0, "DronePlayer", "", 0x7FFFFFFF, 194, "TeleportFog", "TeleportFog");
            TakeInventory("Push Gun",1);
            TakeInventory("Force Gauntlet",1);
            GiveInventory("IsADrone",1);
            tid = defaultTID(-1);
            PlayerTIDs[pln] = tid;
			SetActivator(newTID);
			Thing_SetTranslation(tid,-1);
			SetActivator(tid);
			Thing_Remove(newTID);
          }
        }
        if (isDead(0)) { terminate; }

        GiveInventory("500Health",500);

        delay(1);
      }
      else if (CheckInventory("IsADrone") == 1)
      {
        GiveInventory("PowerPermaFlight",1);
        
        if(DieDieTimeToDie)
        {
          UnmorphActor(0,1);
          TakeInventory("PowerRespawnProtection",1);
          DamageThing(0,MOD_RAILGUN);
          // This is the most brute-force "kill your ass" I've seen.
          // I'm fucking sick and tired of dealing with morphed classes not dying.
          //PrintBold(s:"Player ", d:PlayerNumber(), s:" should be ded");
        }
        else
        {
          /* if(GetCVar("p_drone_buff") && OhShitItsTheFeds)
            { GiveInventory("PushItHarderMakeItBetter"); } */
          GiveInventory("500Health",500);
          TakeInventory("DoomedGuy",1);
        }
        if (isDead(0)) { terminate; }
        delay(1);
      }
    }
}

script PUSH_RESPAWN RESPAWN
{
    ACS_ExecuteAlways(532,0,0,0);
}

script PUSH_DEATH DEATH
{
    //int pln = PlayerNumber();
    //Log(n:pln, s:" was thrown off by ", n:lastShotBy[pln]);

    TakeInventory("ImAlive",1);
    TakeInventory("LightAsAFeather",1);
    TakeInventory("DrawingToolOn",1);
    TakeInventory("DrawingToolReady",1);
    Thing_ChangeTID(0,0);
    PlayerTIDs[PlayerNumber()] = 0;
}

script PUSH_WALLHACK (int which)
{
    int i, time, tid;
    int pln = PlayerNumber();

    switch (which)
    {
      case 1:
        GiveInventory("WallHackVision",1);
        GiveInventory("CannotIntoWallhack", 1);

        while (1)
        {
            for (i = 0; i < PLAYERMAX; i++)
            {
                if (i == pln || !PlayerInGame(i)) { continue; }
                tid = PlayerTIDs[i];
                if (isDead(tid)) { continue; }
                if (!CheckActorInventory(tid,"IsADrone"))
                { HudMessageOnActor(tid, 0x7FFFFFFF, "PLAYMARK", "A", tid); }
            }

            Delay(2);

            if (!CheckInventory("WallhackVision")) { break; }
            TakeInventory("CannotIntoWallhack", 1);

            Delay(2);
            if (!CheckInventory("WallhackVision")) { break; }
        }
        break;

      case 2:
        TakeInventory("WallHackVision",1);
        GiveInventory("CannotIntoWallhack",1);
        Delay(15);
        TakeInventory("CannotIntoWallhack",1);
        break;
    }
}

// The below was made by Kyle873, whom is 7H3 1337357 |-|4(|<3R 0N 7H3 PU5H|\|37

script PUSH_DUNKED (void)
{
    int Time = 52;

    while (Time > 0)
    {
        SetHudSize(320, 240, false);

        // Scanlines
        if ((Timer() % 3) == 0)
            DrawScanlines();

        // Binary
        if ((Timer() % 3) == 0)
            DrawBinary();

        SetFont("bigfont");
        FontX = random(0.00,320.00);
        FontY = random(0.00,240.00);
        hudmessage (s:FontZ[random(0,1)],s:"\n";
        1, 0,FontCL[random(0,2)], FontX, FontY, 0.1);
        delay(1);

        Time--;
        Delay(1);
    }
}

/*<04:55:27> "Kyle873": 
int players = PlayerCount();
int drones = 0;
for (int i = PlayerTIDs; i < PlayerTIDs + MAX_PLAYERS; i++)
    if (CheckActorInventory(i, "IsADrone"))
        drones++;
if (players - drones <= 1)
{
    // We've only got one player left
}

<04:56:25> "Kyle873": 
if (players - drones <= 1)
{
    for (int i = PlayerTIDs; i < PlayerTIDs + MAX_PLAYERS; i++)
        Thing_Damage(i, 1000000;
}